File: solutions/_895_maximum_frequency_stack/solution.py
class FreqStack:

    def __init__(self):
        self.stack = []
        self.freq = {}

    def push(self, val: int) -> None:
        if val in self.freq:
            self.freq[val] += 1
        else:
            self.freq[val] = 1

        if self.freq[val] > len(self.stack):
            self.stack.append([val])
        else:
            self.stack[self.freq[val] - 1].append(val)

    def pop(self) -> int:
        val = self.stack[-1].pop()
        if not self.stack[-1]:
            self.stack.pop()
        self.freq[val] -= 1
        return val


# https://leetcode.com/problems/maximum-frequency-stack/

File: solutions/_895_maximum_frequency_stack/solution.js
class FreqStack {
  constructor() {
    this.stack = [];
    this.freq = {};
  }

  push(val) {
    const count = this.freq[val] + 1 || 1;
    this.freq[val] = count;

    if (count > this.stack.length) {
      this.stack.push([val]);
    } else {
      this.stack[count - 1].push(val);
    }
  }

  pop() {
    const val = this.stack.lastElement().pop();
    if (this.stack.lastElement().isEmpty()) {
      this.stack.pop();
    }
    this.freq[val]--;
    return val;
  }
}

module.exports = { FreqStack };

/*
https://leetcode.com/problems/maximum-frequency-stack/
*/

Array.prototype.isEmpty = function () {
  return this.length === 0;
};

Array.prototype.lastElement = function () {
  return this[this.length - 1];
};